<!-- -*- mode: markdown; coding: utf-8; fill-column: 60; ispell-dictionary: "english" -*- -->

<meta charset="utf-8"/>
<meta name="viewport" content="width=device-width,initial-scale=1"/>
<link rel="stylesheet" href="style.css">


# EDAF75 - lab 1: SQL

**Objective:** to practice writing SQL queries.

This lab is meant to be run as a `jupyter` notebook, you can
download it [here](lab1.zip) (the zip-file contains the
notebook and the database, and a [.html-file](lab1.html)
which you can read in case you have problems reading/running
jupyter notebooks).

## Background

We have a database to handles the academic achievements of
students at LTH -- in it we have three tables:

<hr>
<table rules="all">
 <tr>
  <th align="left"><code>students:</code></th>
  <th></th>
 </tr>
 <tr>
  <td><code>ssn</code></td>
  <td>social security number</td>
 </tr>
 <tr>
  <td><code>first_name</code></td>
  <td>first name (obviously!)</td>
 </tr>
 <tr>
  <td><code>last_name</code></td>
  <td>last name</td>
 </tr>
</table>

<hr>

<table rules="all">
 <tr>
  <th align="left"><code>courses:</code></th>
  <th></th>
 </tr>
 <tr>
  <td><code>course_code</code></td>
  <td>course code, like <code>EDAF75</code></td>
 </tr>
 <tr>
  <td><code>course_name</code></td>
  <td>the name of the course, like "Database Technology"</td>
 </tr>
 <tr>
  <td><code>level</code></td>
  <td>the course level, like "G1", "G2", or "A"</td>
 </tr>
 <tr>
  <td><code>credits</code></td>
  <td>the number of credits for the course, like 7.5</td>
 </tr>
</table>

<hr>

<table rules="all">
 <tr>
  <th align="left"><code>taken_courses:</code></th>
  <th></th>
 </tr>
 <tr>
  <td><code>ssn</code></td>
  <td>the social security number of a student</td>
 </tr>
 <tr>
  <td><code>course_code</code></td>
  <td>the course code for the course the student has taken</td>
 </tr>
 <tr>
  <td><code>grade</code></td>
  <td>the grade for the student passing the course</td>
 </tr>
</table>

<hr>

<center>
  <img src="lab1.png">
</center>

Some sample data:

```text
ssn           first_name   last_name
---           ----------   ---------
861103–2438   Bo           Ek
911212–1746   Eva          Alm
950829–1848   Anna         Nyström
...           ...          ...

course_code   course_name                   level    credits
-----------   -----------                   -----    -------
EDA016        Programmeringsteknik          G1       7.5
EDAA01        Programmeringsteknik - FK     G1       7.5
EDA230        Optimerande kompilatorer      A        7.5
...           ...                           ...      ...

ssn           course_code   grade
---           -----------   -----
861103–2438   EDA016        4
861103–2438   EDAA01        3
911212–1746   EDA016        3
...           ...           ...
```


The tables have been created with the following SQL
statements:

```sql
CREATE TABLE students (
  ssn          CHAR(11),
  first_name   TEXT NOT NULL,
  last_name    TEXT NOT NULL,
  PRIMARY KEY  (ssn)
);

CREATE TABLE courses (
  course_code   CHAR(6),
  course_name   TEXT NOT NULL,
  level         CHAR(2),
  credits       DOUBLE NOT NULL CHECK (credits > 0),
  PRIMARY KEY   (course_code)
);

CREATE TABLE taken_courses (
  ssn           CHAR(11),
  course_code   CHAR(6),
  grade         INTEGER NOT NULL CHECK (grade >= 3 AND grade <= 5),
  PRIMARY KEY   (ssn, course_code),
  FOREIGN KEY   (ssn) REFERENCES students(ssn),
  FOREIGN KEY   (course_code) REFERENCES courses(course_code)
);
```


All courses offered at the "Computer Science and
Engineering" program at LTH during the academic year 2013/14
are in the table 'courses`. Also, the database has been
filled with made up data. SQL statements like the following
have been used to insert the data:

```sql
INTO   students(ssn, first_name, last_name)
VALUES ('950705-2308', 'Anna', 'Johansson'),
       ('930702-3582', 'Anna', 'Johansson'),
       ('911212-1746', 'Eva', 'Alm'),
       ('910707-3787', 'Eva', 'Nilsson'),
       ...
```


## Assignments

As said above, this lab is designed to be run as a jupyter
notebook. If you haven't got jupyter running, you can run
the sql-commands for the problems below interactively inside
sqlite3, or as a script file with all your commands, or
inside some IDE, like [DB Browser for
SQLite](https://sqlitebrowser.org/).

If you're running the lab as a notebook (and hopefully you
are!), evaluate the following cells before you begin:

In [2]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [3]:
%sql sqlite:///lab1.sqlite

u'Connected: @lab1.sqlite'

The tables `students`, `courses` and `taken_courses` are
already in your database, you can see some of their contents
by running the cells below:

In [3]:
%%sql
SELECT  *
FROM    students
LIMIT   4

 * sqlite:///lab1.sqlite
Done.


[(u'950705-2308', u'Anna', u'Johansson'),
 (u'930702-3582', u'Anna', u'Johansson'),
 (u'911212-1746', u'Eva', u'Alm'),
 (u'910707-3787', u'Eva', u'Nilsson')]

In [4]:
%%sql
SELECT  *
FROM    courses
LIMIT   4

 * sqlite:///lab1.sqlite
Done.


[(u'EDA016', u'Programmeringsteknik', u'G1', 7.5),
 (u'EDA031', u'C++ - programmering', u'G2', 7.5),
 (u'EDA040', u'Realtidsprogrammering', u'G2', 6.0),
 (u'EDA050', u'Operativsystem', u'G2', 4.5)]

In [5]:
%%sql
SELECT  *
FROM    taken_courses
LIMIT   4

 * sqlite:///lab1.sqlite
Done.


[(u'950705-2308', u'EITN35', 5),
 (u'950705-2308', u'ESS050', 3),
 (u'950705-2308', u'ETIN70', 4),
 (u'950705-2308', u'FMA140', 4)]

If you inadvertently change the contents of the tables, you
can always recreate the them with the following command (it
must be run at the command line):

```sh
sqlite3 lab1.sqlite < create-lab1-data.sql
```


After some of the questions there is a number in brackets.
This is the number of rows generated by the question. For
instance, [72] after question a) means that there are 72
students in the database, you can use that number as an
indicatation of if your answer is in the right ballpark.

a) What are the names (first name, last name) of all the
   students? [72]

In [6]:
%%sql
SELECT first_name, last_name
FROM students


 * sqlite:///lab1.sqlite
Done.


[(u'Anna', u'Johansson'),
 (u'Anna', u'Johansson'),
 (u'Eva', u'Alm'),
 (u'Eva', u'Nilsson'),
 (u'Elaine', u'Robertson'),
 (u'Maria', u'Nordman'),
 (u'Helena', u'Troberg'),
 (u'Lotta', u'Emanuelsson'),
 (u'Anna', u'Nystr\xf6m'),
 (u'Maria', u'Andersson'),
 (u'Marie', u'Persson'),
 (u'M\xe4rit', u'Aspegren'),
 (u'Ylva', u'Jacobsson'),
 (u'Frida', u'Weidel'),
 (u'Caroline', u'Olsson'),
 (u'Karolin', u'Ek'),
 (u'Eva', u'Hjort'),
 (u'Sofia', u'Kjellberg'),
 (u'Tina', u'L\xf6\xf6f'),
 (u'Lisa', u'Berg'),
 (u'Lotta', u'Bergman'),
 (u'Susanne', u'Dahl'),
 (u'Birgit', u'Ewesson'),
 (u'Ulrika', u'Jonsson'),
 (u'Jenny', u'Lundin'),
 (u'Karin', u'\xd6stberg'),
 (u'Bo', u'Ek'),
 (u'Bo', u'Ek'),
 (u'Bo', u'Ek'),
 (u'Erik', u'Andersson'),
 (u'Erik', u'Andersson'),
 (u'Filip', u'Gustavsson'),
 (u'Niklas', u'Andersson'),
 (u'David', u'Carlsson'),
 (u'Tobias', u'Enberg'),
 (u'Johan', u'Lind'),
 (u'Magnus', u'Pettersson'),
 (u'Filip', u'Persson'),
 (u'Ola', u'Nilsson'),
 (u'Jakob', u'Malmberg'),
 (u'Henrik', u'Berg'),
 (u'John', u'Lind'),
 (u'Patrik', u'Lundh'),
 (u'Mikael', u'Nilsson'),
 (u'Joakim', u'Nilsson'),
 (u'Conny', u'Modig'),
 (u'Christian', u'Wallman'),
 (u'Magnus', u'Str\xf6mgren'),
 (u'Daniel', u'Ahlman'),
 (u'Lars', u'Molin'),
 (u'Magnus', u'Nilsson'),
 (u'Anders', u'Olsson'),
 (u'Martin', u'Alm'),
 (u'H\xe5kan', u'Mossgren'),
 (u'Fredrik', u'Ryd'),
 (u'Mats', u'Wikstr\xf6m'),
 (u'Anders', u'Magnusson'),
 (u'Andreas', u'Molin'),
 (u'Johan', u'Brattberg'),
 (u'Johan', u'Olsson'),
 (u'Jonathan', u'J\xf6nsson'),
 (u'Magnus', u'Hultgren'),
 (u'Axel', u'Nord'),
 (u'Henrik', u'Hult'),
 (u'Joakim', u'Hall'),
 (u'Henrik', u'Gustavsson'),
 (u'Torbj\xf6rn', u'Modig'),
 (u'Daniel', u'Axelsson'),
 (u'Per-Erik', u'Pettersson'),
 (u'Roger', u'Brorsson'),
 (u'Patrik', u'Ohlsson'),
 (u'Peter', u'Solberg')]

b) Same as question a) but produce a sorted listing. Sort
   first by last name and then by first name.

In [9]:
%%sql
SELECT first_name, last_name
FROM students
ORDER BY last_name, first_name


 * sqlite:///lab1.sqlite
Done.


[(u'Daniel', u'Ahlman'),
 (u'Eva', u'Alm'),
 (u'Martin', u'Alm'),
 (u'Erik', u'Andersson'),
 (u'Erik', u'Andersson'),
 (u'Maria', u'Andersson'),
 (u'Niklas', u'Andersson'),
 (u'M\xe4rit', u'Aspegren'),
 (u'Daniel', u'Axelsson'),
 (u'Henrik', u'Berg'),
 (u'Lisa', u'Berg'),
 (u'Lotta', u'Bergman'),
 (u'Johan', u'Brattberg'),
 (u'Roger', u'Brorsson'),
 (u'David', u'Carlsson'),
 (u'Susanne', u'Dahl'),
 (u'Bo', u'Ek'),
 (u'Bo', u'Ek'),
 (u'Bo', u'Ek'),
 (u'Karolin', u'Ek'),
 (u'Lotta', u'Emanuelsson'),
 (u'Tobias', u'Enberg'),
 (u'Birgit', u'Ewesson'),
 (u'Filip', u'Gustavsson'),
 (u'Henrik', u'Gustavsson'),
 (u'Joakim', u'Hall'),
 (u'Eva', u'Hjort'),
 (u'Henrik', u'Hult'),
 (u'Magnus', u'Hultgren'),
 (u'Ylva', u'Jacobsson'),
 (u'Anna', u'Johansson'),
 (u'Anna', u'Johansson'),
 (u'Ulrika', u'Jonsson'),
 (u'Jonathan', u'J\xf6nsson'),
 (u'Sofia', u'Kjellberg'),
 (u'Johan', u'Lind'),
 (u'John', u'Lind'),
 (u'Patrik', u'Lundh'),
 (u'Jenny', u'Lundin'),
 (u'Tina', u'L\xf6\xf6f'),
 (u'Anders', u'Magnusson'),
 (u'Jakob', u'Malmberg'),
 (u'Conny', u'Modig'),
 (u'Torbj\xf6rn', u'Modig'),
 (u'Andreas', u'Molin'),
 (u'Lars', u'Molin'),
 (u'H\xe5kan', u'Mossgren'),
 (u'Eva', u'Nilsson'),
 (u'Joakim', u'Nilsson'),
 (u'Magnus', u'Nilsson'),
 (u'Mikael', u'Nilsson'),
 (u'Ola', u'Nilsson'),
 (u'Axel', u'Nord'),
 (u'Maria', u'Nordman'),
 (u'Anna', u'Nystr\xf6m'),
 (u'Patrik', u'Ohlsson'),
 (u'Anders', u'Olsson'),
 (u'Caroline', u'Olsson'),
 (u'Johan', u'Olsson'),
 (u'Filip', u'Persson'),
 (u'Marie', u'Persson'),
 (u'Magnus', u'Pettersson'),
 (u'Per-Erik', u'Pettersson'),
 (u'Elaine', u'Robertson'),
 (u'Fredrik', u'Ryd'),
 (u'Peter', u'Solberg'),
 (u'Magnus', u'Str\xf6mgren'),
 (u'Helena', u'Troberg'),
 (u'Christian', u'Wallman'),
 (u'Frida', u'Weidel'),
 (u'Mats', u'Wikstr\xf6m'),
 (u'Karin', u'\xd6stberg')]

c) What are the names of the students who were born in 1985?
   [4]

In [18]:
%%sql
SELECT last_name, first_name, ssn
FROM students
WHERE ssn LIKE '85%'


 * sqlite:///lab1.sqlite
Done.


[(u'Jonsson', u'Ulrika', u'850706-2762'),
 (u'Ek', u'Bo', u'850819-2139'),
 (u'Persson', u'Filip', u'850517-2597'),
 (u'Berg', u'Henrik', u'850208-1213')]

d) The next-to-last digit in the social security number is
   even for females, and odd for males. List the first names
   of all female students in our database alphabetically
   (each name should occur only once). Hint: the `substr`
   function can be useful. [26]

In [21]:
%%sql
SELECT last_name, first_name
FROM students
WHERE substr(ssn, 10, 1) % 2 = 0
ORDER BY last_name, first_name


 * sqlite:///lab1.sqlite
Done.


[(u'Alm', u'Eva'),
 (u'Andersson', u'Maria'),
 (u'Aspegren', u'M\xe4rit'),
 (u'Berg', u'Lisa'),
 (u'Bergman', u'Lotta'),
 (u'Dahl', u'Susanne'),
 (u'Ek', u'Karolin'),
 (u'Emanuelsson', u'Lotta'),
 (u'Ewesson', u'Birgit'),
 (u'Hjort', u'Eva'),
 (u'Jacobsson', u'Ylva'),
 (u'Johansson', u'Anna'),
 (u'Johansson', u'Anna'),
 (u'Jonsson', u'Ulrika'),
 (u'Kjellberg', u'Sofia'),
 (u'Lundin', u'Jenny'),
 (u'L\xf6\xf6f', u'Tina'),
 (u'Nilsson', u'Eva'),
 (u'Nordman', u'Maria'),
 (u'Nystr\xf6m', u'Anna'),
 (u'Olsson', u'Caroline'),
 (u'Persson', u'Marie'),
 (u'Robertson', u'Elaine'),
 (u'Troberg', u'Helena'),
 (u'Weidel', u'Frida'),
 (u'\xd6stberg', u'Karin')]

e) How many students are registered in the database?

In [22]:
%%sql
SELECT count() as no_of_students
FROM students


 * sqlite:///lab1.sqlite
Done.


[(72,)]

f) Which courses are offered by the department of
   Mathematics (their course codes have the form `FMAxxx`)?
   [22]

In [25]:
%%sql
SELECT course_name
FROM courses
WHERE course_code like 'FMA%'



 * sqlite:///lab1.sqlite
Done.


[(u'Kontinuerliga system',),
 (u'Optimering',),
 (u'Diskret matematik',),
 (u'Matematiska strukturer',),
 (u'Matristeori',),
 (u'Matristeori, projektdel',),
 (u'Geometri',),
 (u'Olinj\xe4ra dynamiska system',),
 (u'Olinj\xe4ra dynamiska system, projektdel',),
 (u'Bildanalys',),
 (u'Bildanalys, projektdel',),
 (u'Variationskalkyl',),
 (u'Linj\xe4r och kombinatorisk optimering',),
 (u'Datorseende',),
 (u'Datorseende, projektdel',),
 (u'Linj\xe4r algebra',),
 (u'Flerdimensionell analys',),
 (u'Endimensionell analys',),
 (u'Matematik - Funktionsteori',),
 (u'Matematik - System och transformer',),
 (u'Till\xe4mpad matematik - Linj\xe4ra system',),
 (u'Algebraiska strukturer',)]

g) Which courses give more than 7.5 credits? [16]

In [27]:
%%sql
SELECT course_name
FROM courses
WHERE credits > 7.5


 * sqlite:///lab1.sqlite
Done.


[(u'Coachning av programvaruteam',),
 (u'Datorer i system',),
 (u'Till\xe4mpad mekatronik',),
 (u'Mekatronik, industriell produktframtagning',),
 (u'Digitalteknik',),
 (u'Digitala bilder \u2013 kompression',),
 (u'Elektromagnetisk f\xe4ltteori',),
 (u'Elektronik',),
 (u'Introduktionskurs i kinesiska f\xf6r civilingenj\xf6rer',),
 (u'Introduktionskurs i kinesiska f\xf6r civilingenj\xf6rer, del 2',),
 (u'Fysik',),
 (u'Endimensionell analys',),
 (u'Matematisk statistik, allm\xe4n kurs',),
 (u'Realtidssystem',),
 (u'Japanska f\xf6r tekniker',),
 (u'Industriell ekonomi, allm\xe4n kurs',)]

h) How may courses are there for each level (`G1`, `G2`, and
   `A`)?

In [29]:
%%sql
SELECT level, count() AS no_of_courses
FROM courses
GROUP BY level

 * sqlite:///lab1.sqlite
Done.


[(u'A', 87), (u'G1', 31), (u'G2', 60)]

In the next few problems, we'll look at the results of the
student with the social security number 910101–1234 -- to
make things a lot easier, you can create a `VIEW` with all
his results (see lecture 3).

In [39]:
%%sql
DROP VIEW IF EXISTS this_student;
CREATE VIEW this_student AS
SELECT course_code, grade
FROM taken_courses
WHERE ssn is '910101-1234'

 * sqlite:///lab1.sqlite
Done.
Done.


[]

i) Which courses (course codes only) have been taken by the
   student with the social security number 910101–1234? [35]

In [41]:
%%sql
SELECT course_code
FROM this_student

 * sqlite:///lab1.sqlite
Done.


[(u'EDA070',),
 (u'EDA385',),
 (u'EDAA25',),
 (u'EDAF05',),
 (u'EEMN10',),
 (u'EIT020',),
 (u'EIT060',),
 (u'EITF40',),
 (u'EITN40',),
 (u'EITN50',),
 (u'ESS050',),
 (u'ETIF10',),
 (u'ETIF15',),
 (u'ETIN35',),
 (u'ETS052',),
 (u'ETS075',),
 (u'ETS200',),
 (u'ETSA01',),
 (u'EXTA10',),
 (u'EXTA35',),
 (u'FAFF20',),
 (u'FAFF25',),
 (u'FMA145',),
 (u'FMA420',),
 (u'FMAA01',),
 (u'FMAF05',),
 (u'FMS012',),
 (u'FMSF15',),
 (u'FMSN30',),
 (u'FMSN35',),
 (u'FRTN10',),
 (u'GEMA45',),
 (u'GEMA70',),
 (u'MAMF15',),
 (u'TEK210',)]

j) What are the names of these courses, and how many credits
   do they give?

In [42]:
%%sql
SELECT course_code, course_name, credits
FROM this_student
JOIN courses
USING (course_code)


 * sqlite:///lab1.sqlite
Done.


[(u'EDA070', u'Datorer och datoranv\xe4ndning', 3.0),
 (u'EDA385', u'Konstruktion av inbyggda system, f\xf6rdjupningskurs', 7.5),
 (u'EDAA25', u'C-programmering', 3.0),
 (u'EDAF05', u'Algoritmer, datastrukturer och komplexitet', 5.0),
 (u'EEMN10', u'Datorbaserade m\xe4tsystem', 7.5),
 (u'EIT020', u'Digitalteknik', 9.0),
 (u'EIT060', u'Datas\xe4kerhet', 7.5),
 (u'EITF40', u'Digitala och analoga projekt', 7.5),
 (u'EITN40', u'Avancerad webbs\xe4kerhet', 4.0),
 (u'EITN50', u'Avancerad datas\xe4kerhet', 7.5),
 (u'ESS050', u'Elektromagnetisk f\xe4ltteori', 9.0),
 (u'ETIF10', u'Signalbehandling - design och implementering', 7.5),
 (u'ETIF15', u'Medicinsk signalbehandling', 7.5),
 (u'ETIN35', u'IC-projekt 1', 7.5),
 (u'ETS052', u'Datorkommunikation', 4.5),
 (u'ETS075', u'K\xf6system', 4.5),
 (u'ETS200', u'Programvarutestning', 7.5),
 (u'ETSA01', u'Ingenj\xf6rsprocessen f\xf6r programvaruutveckling - metodik', 5.0),
 (u'EXTA10', u'Introduktion till Kinas samh\xe4llsliv, kultur och spr\xe5k', 3.0),
 (u'EXTA35', u'Introduktionskurs i kinesiska f\xf6r civilingenj\xf6rer', 15.0),
 (u'FAFF20', u'Multispektral avbildning', 7.5),
 (u'FAFF25', u'Fysik', 11.0),
 (u'FMA145', u'Olinj\xe4ra dynamiska system, projektdel', 3.0),
 (u'FMA420', u'Linj\xe4r algebra', 6.0),
 (u'FMAA01', u'Endimensionell analys', 15.0),
 (u'FMAF05', u'Matematik - System och transformer', 7.0),
 (u'FMS012', u'Matematisk statistik, allm\xe4n kurs', 9.0),
 (u'FMSF15', u'Markovprocesser', 7.5),
 (u'FMSN30', u'Linj\xe4r och logistisk regression', 7.5),
 (u'FMSN35', u'Station\xe4r och icke-station\xe4r spektralanalys', 7.5),
 (u'FRTN10', u'Flervariabel reglering', 7.5),
 (u'GEMA45', u'F\xf6rst\xe5else och l\xe4rande', 3.0),
 (u'GEMA70', u'Japanska f\xf6r tekniker', 15.0),
 (u'MAMF15', u'Arbetsorganisation och ledarskap', 6.0),
 (u'TEK210', u'Kognition', 4.5)]

k) How many credits has the student taken?

In [43]:
%%sql
SELECT sum(credits) as total_credits
FROM this_student
JOIN courses
USING (course_code)


 * sqlite:///lab1.sqlite
Done.


[(249.5,)]

l) Which is the student’s grade average?

In [46]:
%%sql
SELECT avg(grade) as grade_average
FROM this_student

 * sqlite:///lab1.sqlite
Done.


[(4.0285714285714285,)]

m) Which students have taken 0 credits? [11] This problem
   can be solved in several ways, first do it with an outer
   join:

In [5]:
%%sql
SELECT DISTINCT first_name, last_name
FROM students
LEFT JOIN taken_courses
USING (ssn)
WHERE taken_courses.ssn is null

 * sqlite:///lab1.sqlite
Done.


[(u'Anna', u'Nystr\xf6m'),
 (u'Caroline', u'Olsson'),
 (u'Bo', u'Ek'),
 (u'Erik', u'Andersson'),
 (u'Johan', u'Lind'),
 (u'Filip', u'Persson'),
 (u'Jonathan', u'J\xf6nsson'),
 (u'Magnus', u'Hultgren'),
 (u'Joakim', u'Hall'),
 (u'Henrik', u'Gustavsson')]

Now do the same thing using a subquery:

In [56]:
%%sql
SELECT first_name, last_name
FROM students
WHERE ssn NOT IN (
    SELECT ssn
    FROM taken_courses
                    )

 * sqlite:///lab1.sqlite
Done.


[(u'Anna', u'Nystr\xf6m'),
 (u'Caroline', u'Olsson'),
 (u'Bo', u'Ek'),
 (u'Erik', u'Andersson'),
 (u'Erik', u'Andersson'),
 (u'Johan', u'Lind'),
 (u'Filip', u'Persson'),
 (u'Jonathan', u'J\xf6nsson'),
 (u'Magnus', u'Hultgren'),
 (u'Joakim', u'Hall'),
 (u'Henrik', u'Gustavsson')]

n) List the names and average grades of the 10 students with
   the highest grade average?

In [6]:
%%sql
SELECT first_name, last_name, round(avg(grade), 3) AS average_grade
FROM students
JOIN taken_courses
USING (ssn)
GROUP BY (ssn)
ORDER BY avg(grade) DESC
LIMIT 10


 * sqlite:///lab1.sqlite
Done.


[(u'Bo', u'Ek', 4.35),
 (u'Helena', u'Troberg', 4.308),
 (u'Elaine', u'Robertson', 4.235),
 (u'Anna', u'Johansson', 4.231),
 (u'Ylva', u'Jacobsson', 4.219),
 (u'Anna', u'Johansson', 4.2),
 (u'Mikael', u'Nilsson', 4.174),
 (u'Jakob', u'Malmberg', 4.167),
 (u'Maria', u'Andersson', 4.158),
 (u'Per-Erik', u'Pettersson', 4.154)]

o) List the social security number and total number of
   credits for all students. Students with no credits should
   be included in the listing, with 0 credits (not `NULL`).

   If you do this with an outer join (and I recommend you
   try it!) you might want to use the function
   `coalesce(v1, v2, ...)`; it returns the first value which
   is not `NULL`, so `coalesce(avg(grade), 0)` would give 0
   if the were no grades (i.e., if `grade` were `NULL`), you
   can also try the `ifnull` function.

   (It may be a little bit tricky to get this query right,
   if you're missing the students with 0 credits, don't
   worry, your TA will help you get it right). [72]

In [11]:
%%sql

SELECT ssn, coalesce(sum(grade), 0)
FROM students
LEFT JOIN taken_courses
USING (ssn)
Left JOIN courses
using (course_code)
GROUP BY ssn
ORDER BY credits

 * sqlite:///lab1.sqlite
Done.


[(u'850517-2597', 0),
 (u'870909-3367', 0),
 (u'880206-1915', 0),
 (u'891007-3091', 0),
 (u'891220-1393', 0),
 (u'900129-3374', 0),
 (u'900313-2257', 0),
 (u'911015-3758', 0),
 (u'931225-3158', 0),
 (u'950125-1153', 0),
 (u'950829-1848', 0),
 (u'930420-2833', 4),
 (u'880620-2564', 199),
 (u'881110-1272', 78),
 (u'891021-1287', 143),
 (u'900528-1540', 39),
 (u'910101-1234', 141),
 (u'940825-1934', 62),
 (u'860323-1071', 41),
 (u'910707-3787', 13),
 (u'920514-3433', 15),
 (u'930502-2040', 110),
 (u'931128-1333', 95),
 (u'850208-1213', 82),
 (u'850706-2762', 15),
 (u'850819-2139', 48),
 (u'860206-1065', 89),
 (u'860819-2864', 79),
 (u'870915-2742', 145),
 (u'880322-3951', 141),
 (u'881030-2772', 74),
 (u'890103-1256', 31),
 (u'890622-1928', 116),
 (u'890820-2496', 99),
 (u'891231-2554', 47),
 (u'900206-3735', 83),
 (u'900526-1450', 16),
 (u'900918-3999', 128),
 (u'901030-1895', 108),
 (u'910226-3409', 37),
 (u'910308-1826', 56),
 (u'910915-2068', 199),
 (u'911020-3538', 140),
 (u'911212-1746', 103),
 (u'920308-3854', 170),
 (u'920623-3258', 184),
 (u'920812-1857', 25),
 (u'920906-3371', 97),
 (u'920921-2499', 149),
 (u'921029-1995', 143),
 (u'921222-2113', 187),
 (u'930702-3582', 165),
 (u'930716-3933', 28),
 (u'930804-3050', 86),
 (u'931015-1271', 12),
 (u'931208-3605', 135),
 (u'931213-2824', 72),
 (u'940318-2373', 130),
 (u'940801-2971', 96),
 (u'940802-3861', 76),
 (u'941003-1225', 25),
 (u'950317-3520', 24),
 (u'950705-2308', 21),
 (u'950714-2190', 18),
 (u'950715-2090', 4),
 (u'950830-2637', 92),
 (u'951004-2346', 189),
 (u'951122-1048', 55),
 (u'861103-2438', 87),
 (u'890621-3057', 170),
 (u'891106-1277', 30),
 (u'900227-2369', 148)]

p) Is there more than one student with the same name? If so,
   who are these students and what are their social security
   numbers? [7]

   As usual there are several ways of solving this, try to
   solve it using a `WITH`-statement where you create a
   table with all duplicate names, and then use it in a
   `JOIN`:

In [72]:
%%sql
WITH duplicates AS(
    SELECT first_name, last_name
    FROM students
    GROUP BY first_name, last_name
    HAVING count() > 1
)

SELECT first_name, last_name, ssn
FROM students
JOIN duplicates
USING (first_name, last_name)

 * sqlite:///lab1.sqlite
Done.


[(u'Anna', u'Johansson', u'950705-2308'),
 (u'Anna', u'Johansson', u'930702-3582'),
 (u'Bo', u'Ek', u'861103-2438'),
 (u'Bo', u'Ek', u'931225-3158'),
 (u'Bo', u'Ek', u'850819-2139'),
 (u'Erik', u'Andersson', u'891220-1393'),
 (u'Erik', u'Andersson', u'900313-2257')]

q) What 5 courses have the highest grade average?

In [7]:
%%sql
SELECT course_name, round(avg(grade), 3) AS grade_average
FROM courses
JOIN taken_courses
USING (course_code)
GROUP BY course_code



 * sqlite:///lab1.sqlite
Done.


[(u'Programmeringsteknik', 3.8),
 (u'C++ - programmering', 3.833),
 (u'Realtidsprogrammering', 3.667),
 (u'Operativsystem', 3.5),
 (u'Datorer och datoranv\xe4ndning', 3.6),
 (u'N\xe4tverksprogrammering', 4.5),
 (u'Till\xe4mpad artificiell intelligens', 3.8),
 (u'Kompilatorteknik', 4.222),
 (u'Datorgrafik', 4.333),
 (u'Optimerande kompilatorer', 4.286),
 (u'Programvaruutveckling i grupp \u2013 projekt', 4.222),
 (u'Coachning av programvaruteam', 3.5),
 (u'Konstruktion av inbyggda system, f\xf6rdjupningskurs', 4.143),
 (u'Programmeringsteknik - f\xf6rdjupningskurs', 3.8),
 (u'Datorer i system', 4.2),
 (u'C-programmering', 3.667),
 (u'Operativsystem - projekt', 4.25),
 (u'Algoritmer, datastrukturer och komplexitet', 3.857),
 (u'Objektorienterad modellering och diskreta strukturer', 4.182),
 (u'Algoritmimplementering', 4.167),
 (u'Databasteknik', 3.75),
 (u'Constraint-programmering', 3.889),
 (u'Konfigurationshantering', 3.889),
 (u'Konstruktion av inbyggda system', 4.111),
 (u'Spr\xe5kteknologi', 4.0),
 (u'Multicoreprogrammering', 3.5),
 (u'Fotorealistisk datorgrafik', 3.714),
 (u'H\xf6gpresterande datorgrafik', 4.0),
 (u'Funktionsprogrammering', 4.167),
 (u'Intelligenta system - projekt', 4.5),
 (u'Avancerade algoritmer', 3.571),
 (u'Spr\xe5kteknologi: Projekt', 3.875),
 (u'Kodningsteknik', 3.667),
 (u'Kryptoteknik', 3.833),
 (u'Matematisk kryptologi', 3.9),
 (u'Sensorteknik', 3.5),
 (u'Medicinsk m\xe4tteknik', 3.833),
 (u'Mikrosensorer', 3.6),
 (u'EMC, st\xf6rningar och st\xf6rningsbegr\xe4nsning', 3.0),
 (u'Datorbaserade m\xe4tsystem', 4.0),
 (u'Ultraljudsfysik och teknik', 3.833),
 (u'Projekt i industriell elektroteknik och automation', 4.25),
 (u'Till\xe4mpad mekatronik', 3.9),
 (u'Mekatronik, industriell produktframtagning', 4.375),
 (u'Digitalteknik', 4.143),
 (u'Datas\xe4kerhet', 4.167),
 (u'Datorteknik', 4.0),
 (u'OFDM f\xf6r bredbandskommunikation', 4.0),
 (u'Digitala bilder \u2013 kompression', 4.2),
 (u'Webbs\xe4kerhet', 3.8),
 (u'Datorarkitektur', 3.857),
 (u'Digitala strukturer p\xe5 kisel', 4.333),
 (u'Digitala och analoga projekt', 4.75),
 (u'Multipelantennsystem', 4.125),
 (u'Internet inuti', 3.75),
 (u'Projekt i elektro- och informationsteknik', 4.429),
 (u'Avancerad webbs\xe4kerhet', 4.143),
 (u'Informationsteori', 3.667),
 (u'Avancerad datas\xe4kerhet', 4.429),
 (u'Komponentfysik', 4.375),
 (u'Elektromagnetisk f\xe4ltteori', 3.75),
 (u'M\xe4tteknik', 4.125),
 (u'Signalbehandling i multimedia', 4.6),
 (u'Elektronik', 3.875),
 (u'Patent och annan immaterialr\xe4tt', 4.333),
 (u'Grundl\xe4ggande radioteknik', 3.625),
 (u'Signalbehandling - design och implementering', 3.857),
 (u'Medicinsk signalbehandling', 4.571),
 (u'Kanalmodellering f\xf6r tr\xe5dl\xf6s kommunikation', 4.286),
 (u'Radiosystem', 4.2),
 (u'Digital IC-konstruktion', 4.167),
 (u'IC-projekt 1', 4.333),
 (u'IC-projekt 2', 4.0),
 (u'DSP-design', 4.0),
 (u'Integrerade A/D och D/A omvandlare', 4.4),
 (u'Avancerade A/D och D/A omvandlare', 3.556),
 (u'Modern elektronik', 4.0),
 (u'Algoritmer i signalprocessorer - projektkurs', 3.0),
 (u'Datorkommunikation', 3.556),
 (u'Simulering', 3.75),
 (u'K\xf6system', 4.0),
 (u'Kravhantering', 4.0),
 (u'Programvarutestning', 4.4),
 (u'Ingenj\xf6rsprocessen f\xf6r programvaruutveckling - metodik', 4.143),
 (u'Ingenj\xf6rsprocessen f\xf6r programvaruutveckling - samh\xe4llsaspekter', 3.917),
 (u'Ingenj\xf6rsprocessen f\xf6r programvaruutveckling - ekonomi och kvalitet', 3.889),
 (u'Internetprotokoll', 4.222),
 (u'Avancerad telekommunikation', 4.111),
 (u'Programvaruutveckling f\xf6r stora system', 3.5),
 (u'Digital kommunikation', 4.2),
 (u'Digital kommunikation, forts\xe4ttningskurs', 3.4),
 (u'Adaptiv signalbehandling', 4.167),
 (u'Optimal signalbehandling', 3.429),
 (u'Moderna tr\xe5dl\xf6sa system - LTE och dess efterf\xf6ljare', 4.4),
 (u'Introduktion till Kinas samh\xe4llsliv, kultur och spr\xe5k', 3.727),
 (u'Introduktionskurs i kinesiska f\xf6r civilingenj\xf6rer', 3.857),
 (u'Introduktionskurs i kinesiska f\

r) (Not required) What are the 'best' initial letters of the
   last names, i.e., if you take the average grades for each
   first letter of the last name, which three initials have
   the highest averages? That is, we want a table like (but
   probably with other values):

<center>
<table rules="all">
  <tr>
   <th>initial</th>
   <th>gpa</th>
  </tr>
  <tr>
   <td>A</td>
   <td>4.2</td>
  </tr>
  <tr>
   <td>L</td>
   <td>4.18</td>
  </tr>
  <tr>
   <td>E</td>
   <td>4.0</td>
  </tr>
</table>
</center>

In [ ]:
%%sql
